### ISSUE TITLE
Low GPU utilization with NEST and OOMptimizer

### ISSUE THREAD

**Issue by MahmoudAshraf97** [NVIDIA-NeMo/NeMo#13619]
Hi,
I'm using the default NEST config with a dataset that is around 9k hours, I use tarred dataset and OOMptimizer with a single H100, but the utilization is lower than expected, This is my training dataset config
```yaml
train_ds:
    manifest_filepath: /home/mahmo/development/data/tarred_everything/tarred_audio_manifest.json
    noise_manifest: null
    sample_rate: 16000
    batch_size: 512
    batch_duration: 1100
    shuffle: true
    num_workers: 8
    pin_memory: true
    use_start_end_token: false
    trim_silence: false
    max_duration: 60.0
    min_duration: 1.0
    drop_last: true
    use_lhotse: true
    is_tarred: true
    tarred_audio_filepaths: /home/mahmo/development/data/tarred_everything/audio_0.tar
    use_bucketing: true
    bucket_duration_bins: [10.08325,13.92325,17.076,19.448,19.744,19.9,19.968,20.0,20.064,20.224,20.48,24.069,29.91019,30.0]
    bucket_batch_size: [238,167,135,115,113,112,112,111,111,110,108,89,69,68]
    batch_augmentor:
      _target_: nemo.collections.asr.modules.ssl_modules.MultiSpeakerNoiseAugmentation
      prob: 0.2
      noise_ratio: 0.5
      min_r_speech: -5.0
      max_r_speech: 5.0
      min_r_noise: -5.0
      max_r_noise: 20.0
      min_mix_rate: 0.5
      max_mix_rate: 0.5
      min_num_segments: 1
      max_num_segments: 1
      min_num_speakers: 1
      max_num_speakers: 1
```
These are the metrics on wandb:

![Image](https://github.com/user-attachments/assets/83f3edae-e2c4-41b4-908a-322843460b6e)

![Image](https://github.com/user-attachments/assets/76c2a63a-2717-4153-a69e-f3438fa8a638)

Note that with regular Fast conformer with CTC loss I'm able to get consistent ~95% utilization with a batch size of 32 only along with lhotse and tarred dataset
cc: @pzelasko @stevehuang52 
Thanks in advance

**Comment by MahmoudAshraf97** [comment]
This is another experiment with 8x H100, I kept everything the same as above except for using a regular sharded dataset instead of a tarred dataset, and divided the batch sizes by 8 to keep the global batch size the same

![Image](https://github.com/user-attachments/assets/3350484e-74e0-496b-b488-757c565de7e3)

Supposedly in this configuration, I'm using a slower dataloader and a smaller batch size which in general should lead to lower utilization, but it's the same

**Comment by MahmoudAshraf97** [comment]
not stale

**Comment by stevehuang52** [comment]
@pzelasko Do you have suggestion on this?

I'm not sure about the root cause, but high GPU memory usage doesn't always mean high GPU utilization. The large memory consumption may come from the large embedding layer, which doesn't contain a lot of math operations. 

One thing to try is using mixed precision (`trainer.precision=bf16-mixed`), which will reduce the memory usage so that you can further increase the batch size, which will lead to higher compute usage.

Another thing is to set `batch_size: null` in your config, so that lhotse will only look at the batch durations.

**Comment by MahmoudAshraf97** [comment]
I'm preparing another training run with the following things changed, Hopefully I can get better utilization:
1. Use a RAID0 Array to host the tarred dataset which has 6.5GB/s read speed instead of the 1GB/s disk used above
2. Shard the dataset to 1024 shards, the above experiment was done with only one shard


btw I'm using `bf16-mixed` already, I tried playing with the batch size but the utilization is still the same

**Comment by MahmoudAshraf97** [comment]
This is the configuration for the new run:
```yaml
train_ds:
    manifest_filepath: "/mnt/tarred_dataset/dataset/sharded_manifests/manifest_{0..99}.json"
    tarred_audio_filepaths: "/mnt/tarred_dataset/dataset/audio_{0..99}.tar"
    noise_manifest: null
    sample_rate: 16000
    batch_size: null
    batch_duration: 5000
    shuffle: true
    num_workers: 8
    pin_memory: true
    use_start_end_token: false
    trim_silence: false
    max_duration: 60.0
    min_duration: 0.0
    drop_last: true
    is_concat: false
    concat_sampling_technique: temperature
    concat_sampling_temperature: 1.0
    use_lhotse: true
    is_tarred: true
    use_bucketing: true
    bucket_duration_bins: [11,12,13,14,15,16,17,18,19,20,21,22,23,24]
    bucket_batch_size: [217,195,181,166,154,142,133,126,118,110,104,99,94,84]
    batch_augmentor:
      _target_: nemo.collections.asr.modules.ssl_modules.MultiSpeakerNoiseAugmentation
      prob: 0.2
      noise_ratio: 0.5
      min_r_speech: -5.0
      max_r_speech: 5.0
      min_r_noise: -5.0
      max_r_noise: 20.0
      min_mix_rate: 0.5
      max_mix_rate: 0.5
      min_num_segments: 1
      max_num_segments: 1
      min_num_speakers: 1
      max_num_speakers: 1
```

The utilization is still very low using 8xH100 machine, note that the average read usage of the SSD array is 100MB/s as per `iostat`, it's capable of doing 6,500 MB/s

![Image](https://github.com/user-attachments/assets/f5a559ac-1614-4d16-b7c0-241d257c5f51)

**Comment by MahmoudAshraf97** [comment]
I managed to narrow down the problem to the data loader, by creating a dummy sample and running a training loop that only uses this sample, the utilization is around 80-85% when masking is enabled and a constant 96% when masking is disabled

when using the dataloader, the utilization follows the pattern above, the dataloader config is the same as above

**Comment by MahmoudAshraf97** [comment]
For comparison, I set up two dataloaders with identical simple configs:
```python
model.cfg.train_ds.manifest_filepath = "manifest_{0..99}.json"
model.cfg.train_ds.tarred_audio_filepaths = "audio_{0..99}.tar"
model.cfg.train_ds.batch_size = 128
model.cfg.train_ds.use_bucketing=False
model.cfg.train_ds.num_workers=8
```
The difference is the two model classes are `EncDecDenoiseMaskedTokenPredModel` which uses `LhotseAudioNoiseDataset` and `EncDecCTCModelBPE` which uses `LhotseSpeechToTextBpeDataset`
the SSL model dataloader loads a sample in an average of 116 ms while the CTC model dataloader takes only 25 ms, increasing the number of workers does not make a difference
I guess the difference in speed explains the low utilization
cc @pzelasko

**Comment by MahmoudAshraf97** [comment]
still active
